ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,428 @@
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTarget
+metadata:
+    name: prod-dt
+spec:
+    deploymentTargetClassName: isolation-level-namespace
+    kubernetesCredentials:
+        defaultNamespace: team-a--prod-dtc
+        apiURL: …
+        clusterCredentialsSecret: team-a--prod-dtc--secret
+    claimRef:
+        name: prod-dtc
+```
+
+#### DeploymentTargetClaim (DTC)
+
+Represents a request for a [DeploymentTarget]. The phase field indicates if there is a [DT] that
+fulfills the requests of the [DTC] and whether it has been bound to it.
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClaim
+metadata:
+    name: prod-dtc
+spec:
+    deploymentTargetClassName: isolation-level-namespace
+status:
+    phase: Bound
+```
+
+#### DeploymentTargetClass (DTCLS)
+
+Referred from a [DeploymentTarget] and [DeploymentTargetClaim]. Defines [DeploymentTarget] properties that
+should be abstracted from the controller/user that creates a [DTC] and wants a [DT] to be provisioned
+automatically for it.
+
+In the example below you can see a class that represents a [DT] that grants the requestor access to
+a namespace. The requestor isn’t aware of the actual location the [DT] is going to be provisioned. The
+parameters section can be used to forward additional information to the provisioner. The
+reclaimPolicy field will tell the provisioner what to do with the [DT] once its corresponding [DTC] is
+deleted, the values can be Retain or Delete.
+
+**Immutable object**: yes
+
+**Scope**: cluster
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClass
+metadata:
+    name: isolation-level-namespace
+spec:
+    provisioner: appstudio.redhat.com/devsandbox
+    parameters: {}
+    reclaimPolicy: Delete
+```
+
+#### Environment
+
+[Environment] objects refer to a [DTC] using the `deploymentTargetClaim`
+field. The environment controller will wait for the [DTC] to get to the
+“bound” phase, once it is bound, it will reach the [DT] and read the
+target's connection details from the kubernetesCredentials field and
+configure Argo/Gitops services to use them.
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: Environment
+metadata:
+    name: prod
+spec:
+    displayName: “Production for Team A”
+    deploymentStrategy: AppStudioAutomated
+    parentEnvironment: staging
+    tags: [prod]
+    configuration:
+        target:
+            deploymentTargetClaim:
+                claimName: prod-dtc
+```
+
+### Controllers
+
+#### DeploymentTargetBinder
+
+Binds [DeploymentTargetClaim] to a [DeploymentTarget] that satisfies its requirements.
+
+It watches for the creation of new [DTC] and tries to find a matching [DT] for each one of them.
+
+A [DT] that we created dynamically for a specific [DTC] will always be attached to it.
+
+[DT] and [DTC] have one to one bindings.
```

Comment:
What would this look like if we had multiple workspaces using the same cluster? Would you create multiple deployment targets for the same cluster?
And if we had multiple environments using the same cluster within a workspace, would we also have multiple deployment targets for that cluster?

Reply:
In terms of the user's appstudio namespace where the DT and the DTC live, there is also a Secret that contains the kubeconfig-formatted serviceaccount token to admin the deployment namespace. It is no more persistent than an ordinary Secret in k8s.